src/groupeddataframe/grouping.jl

```diff
@@ -245,39 +337,110 @@ end
 
-function _combine(first::Union{NamedTuple, DataFrameRow}, f::Function,
+function _combine(f::Union{AbstractVector{<:Pair}, Tuple{Vararg{Pair}},
+                           NamedTuple{<:Any, <:Tuple{Vararg{Pair}}}},
                   gd::GroupedDataFrame)
-    m = length(first)
-    n = length(gd)
-    idx = Vector{Int}(undef, n)
-    initialcols = ntuple(i -> Tables.allocatecolumn(typeof(first[i]), n), m)
-    cols = _combine!(first, initialcols, idx, 1, 1, f, gd, tuple(propertynames(first)...))
-    valscat = DataFrame(collect(cols), collect(propertynames(first)))
+    idxs = (first(p) isa Union{Integer, Symbol} ?
+            index(gd.parent)[first(p)] :
+            index(gd.parent)[collect(first(p))] for p in f)
+    incols = [names(gd.parent)[idx] for idx in idxs]
+    allcols = collect(reduce(union, (nam isa Symbol ? (nam,) : nam for nam in incols)))
+    f2 = ntuple(length(f)) do i
+        nms = incols[i]
+        (nms isa AbstractArray ? Tuple(nms) : nms) => last(f[i])
+    end
+    # Use temporary names for columns, and rename after the fact where appropriate
+    tmpnms = Tuple(gennames(length(f)))::NTuple{length(f), Symbol}
+    fun = function(x)
+        tup = map(f2) do p
+            local incols = first(p)
+            res = incols isa Tuple ?
+                  last(p)(NamedTuple{incols}(map(c -> x[c], incols))) :
+                  last(p)(x[incols])
+            if res isa Union{AbstractDataFrame, NamedTuple, DataFrameRow, AbstractVector, AbstractMatrix}
+                throw(ArgumentError("a single value result is required when passing a vector or tuple " *
+                                    "of functions (got $(typeof(res)))"))
+            end
+            res
+        end
+        NamedTuple{tmpnms}(tup)
+    end
+    idx, valscat = _combine(allcols => fun, gd)
+    if f isa NamedTuple
+        nams = collect(propertynames(f))
+    else
+        nams = names(valscat)
+        for i in 1:ncol(valscat)
+            if f[i] isa Pair{<:Union{Symbol,Integer}}
+                 nams[i] = Symbol(names(gd.parent)[index(gd.parent)[first(f[i])]],
+                                  '_', funname(last(f[i])))
+            end
+        end
+    end
+    names!(valscat, nams)
     idx, valscat
 end
 
-function _combine(first::AbstractDataFrame, f::Function, gd::GroupedDataFrame)
-    m = size(first, 2)
-    idx = Vector{Int}()
-    initialcols = ntuple(i -> similar(first[i], 0), m)
-    cols = _combine!(first, initialcols, idx, 1, 1, f, gd, names(first))
-    valscat = DataFrame(collect(cols), names(first))
+function _combine(f::Any, gd::GroupedDataFrame)
+    if f isa Pair{<:Union{Symbol,Integer}}
+        incols = (gd.parent[first(f)],)
+        fun = last(f)
+    elseif f isa Pair
+        df = gd.parent[collect(first(f))]
+        incols = NamedTuple{Tuple(names(df))}(columns(df))
+        fun = last(f)
+    else
+        incols = nothing
+        fun = f
+    end
+    firstres = do_call(fun, gd, incols, 1)
```

Comment:
OK - I see here that you use the `Tuple` version of `do_call` only when tuple has one element - right?

Reply:
This can be marked as resolved - right?